pkg/codegen/schema/loader.go

```diff
@@ -224,7 +224,7 @@ func (l *pluginLoader) loadSchemaBytes(pkg string, version *semver.Version) ([]b
 	}
 
 	// don't ues the cache if there is a valid version
-	if pluginInfo.SchemaPath != "" && version == nil {
+	if pluginInfo.SchemaPath != "" {
 		schemaBytes, ok := l.loadCachedSchemaBytes(pkg, pluginInfo.SchemaPath, pluginInfo.SchemaTime)
```

Comment:
Can we solve this by encoding path into `SchemaPath` (as set [here](https://github.com/pulumi/pulumi/blob/c082382b04902c82146fcb7b7f4734f80c1d9370/sdk/go/common/workspace/plugins.go#L675))?

I don't want to retrieve a cached schema with the wrong version.

Reply:
I think the check might just be reversed? Should it be `version != nil`?

Reply:
I've fixed up the first commit to remove the comment (redundant) and flip the condition.

Reply:
I think the check is because our caching isn't aware of versions. What we need is a cache key that looks like `(name, version)`. If version is `nil`, I think we can allow that to degenerate to `(name, *)`.

Reply:
It is important that `loadSchemaBytes("aws", "4.37.0")` doesn't retrieve the cached result of `loadSchemaBytes("aws", "5.3.0")`. I think thats what the check was there for.

Reply:
The `deploytest.PluginLoader` uses the schemas hardcoded in the makefile (`schema-aws!4.26.0` -> `aws.json`). The issue wasn't that `loadSchemaBytes` did the wrong thing, it's that the loader returned a schema path that wasn't valid for caching because it ignored versions:

https://github.com/pulumi/pulumi/blob/d93383b235482e63c5a538d583e837eb3d659991/pkg/resource/deploy/deploytest/pluginhost.go#L400-L403

Reply:
Maybe I'm missing something, but I think this will effect more than the loader we use during tests. Non-test codegen doesn't rely on the `deploy test.PluginLoader`.

Reply:
Looking at [this change](https://github.com/pulumi/pulumi/blob/aaf6cbcc4fce59aa062acd6adf311bb03a466c87/pkg/resource/deploy/deploytest/pluginhost.go#L413), I think we're good here.